httpconnect/httpconnect.go

```diff
@@ -147,28 +151,37 @@ func (f *HTTPConnectHandler) intercept(w http.ResponseWriter, req *http.Request)
 			}
 		}
 	}
-	var closeOnce sync.Once
-	go func() {
+
+	readErrCh := make(chan error)
+	op.Go(func() {
 		buf := buffers.Get()
 		defer buffers.Put(buf)
-		if _, err := io.CopyBuffer(connOut, clientConn, buf); err != nil {
-			log.Debug(err)
+		_, readErr := io.CopyBuffer(connOut, clientConn, buf)
+		if readErr != nil {
+			log.Debug(readErr)
 		}
-		closeOnce.Do(closeConns)
-
-	}()
+		readErrCh <- readErr
+	})
 	buf := buffers.Get()
 	defer buffers.Put(buf)
-	if _, err := io.CopyBuffer(clientConn, connOut, buf); err != nil {
-		log.Debug(err)
```

Comment:
It's interesting that you mention the error case.  I was thinking more about the success case, wherein we finish reading before writing.  In that case, the write might fail because we closed the connection already.

In the case of an error, perhaps it's good to close the connection right away to deliberately interrupt unfinished writes?